train_decoder.py

```diff
@@ -0,0 +1,495 @@
+        # Then we are recalling from wandb
+        print(f"Recalling trainer from wandb: {wandb_run_path}/{wandb_file_path}")
+        trainer_state_dict_file = wandb.restore(wandb_file_path, run_path=wandb_run_path)
+        state_dict = torch.load(trainer_state_dict_file.name)
+    else:
+        # Then we are recalling from a file
+        print(f"Recalling trainer from file: {local_filepath}")
+        state_dict = torch.load(local_filepath)
+    apply_trainer_state_dict(trainer, state_dict)
+    return trainer, state_dict["epoch"], state_dict["step"], state_dict["validation_losses"] if "validation_losses" in state_dict else []
+
+def train(
+    dataloaders,
+    decoder,
+    inference_device,
+    using_wandb=False,
+    resume_from=None,
+    resume_config=None,
+    epoch_samples = None,  # If the training dataset is resampling, we have to manually stop an epoch
+    validation_samples = None,
+    epochs = 20,
+    n_sample_images = 5,
+    save_every_n_samples = 100000,
+    amp=False,
+    lr=1e-4,
+    wd=0,
+    use_ema=True,
+    max_grad_norm=None,
+    base_path=None,
+    save_all=False,
+    save_latest=True,
+    save_best=True,
+    **kwargs
+):
+    """
+    Trains a decoder on a dataset.
+    """
+    # Create the base path if it doesn't exist
+    if not os.path.exists(base_path):
+        os.makedirs(base_path)
+    # If we are saving all, we also need to make a checkpoints folder inside there
+    if save_all:
+        if not os.path.exists(os.path.join(base_path, "checkpoints")):
+            os.makedirs(os.path.join(base_path, "checkpoints"))
+    trainer = DecoderTrainer(
+        decoder,
+        lr = lr,
+        wd = wd,
+        amp = amp,
+        use_ema = use_ema,
+        max_grad_norm = max_grad_norm
+    )
+    # Set up starting model and parameters based on a recalled state dict
+    start_step = 0
+    start_epoch = 0
+    validation_losses = []
+    if resume_from is not None:
+        if resume_from == "wandb":
+            trainer, start_epoch, start_step, validation_losses = recall_trainer(
+                trainer,
+                wandb_run_path=resume_config["wandb_run_path"],
+                wandb_file_path=resume_config["wandb_file_path"],
+            )
+        elif resume_from == "file":
+            trainer, start_epoch, start_step, validation_losses = recall_trainer(
+                trainer,
+                local_filepath=resume_config["filepath"]
+            )
+        else:
+            raise ValueError("resume_from must be either wandb or file")
+    trainer = trainer.to(device=inference_device)
+    
+    send_to_device = lambda arr: [x.to(device=inference_device, dtype=torch.float) for x in arr]
+    step = start_step
+    for epoch in range(start_epoch, epochs):
+        print(print_ribon(f"Starting epoch {epoch}", repeat=40))
+        trainer.train()
+        decoder.train()  # I don't see anything in the trainer that would set decoder to train mode so I assume we also need this
```

Comment:
oh no, you won't need this, because the trainer is also a `nn.Module` class, so it will propagate the train signal to all of its children

Reply:
I need some documentation on how `to`, `train`/`eval`, and `state_dict` propagate. I assumed at first that they all propagate, but that was wrong because `state_dict` doesn't at least. Then an error resolved itself when I separately sent the trainer and decoder to the device, but I guess that was coincidental and `to` does actually propagate.

I'll make a note to go through and remove calls to anything that may be propagating.